### PR TITLE
Adjust snowpark build story

### PR DESCRIPTION
**Pull request by sfc-gh-turbaszek**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
snowpark build and deploy commands use information from project definition to build and upload the artefact.

### CODE REVIEW

File: src/snowcli/cli/snowpark/commands.py [hunk 1]
```diff
@@ -165,34 +207,43 @@ def deploy(
...
     packages: List[str],
-    stage_manager: StageManager,
-    build_artifact_path: Path,
-    install_coverage_wrapper: bool = False,
+    stage_artifact_path: str,
 ):
     identifier = build_udf_sproc_identifier(object_definition)
     log.info(f"Deploying {object_type}: {identifier}")
-    handler = object_definition["handler"]
+    handler = f'{object_definition["handler"]}'
```
Comment (sfc-gh-turbaszek) [inline src/snowcli/cli/snowpark/commands.py:242]:
```suggestion
    handler = object_definition["handler"]
```

File: src/snowcli/cli/snowpark/commands.py [hunk 2]
```diff
@@ -261,12 +306,19 @@ def build(
...
-        check_anaconda_for_pypi_deps,
-        package_native_libraries,  # type: ignore[arg-type]
+        source=source,
+        artefact_file=artefact_file,
+        pypi_download=pypi_download,  # type: ignore[arg-type]
+        check_anaconda_for_pypi_deps=check_anaconda_for_pypi_deps,
+        package_native_libraries=package_native_libraries,  # type: ignore[arg-type]
     )
-    return MessageResult("Done")
+    return MessageResult(f"Build done. Artifact path: {artefact_file}")
```
Comment (sfc-gh-turbaszek) [inline src/snowcli/cli/snowpark/commands.py:321]:
```suggestion
    return MessageResult(f"Build done. Artefact path: {artefact_file}")
```

File: src/snowcli/zipper.py [hunk 3]
```diff
@@ -34,34 +39,24 @@ def add_file_to_existing_zip(zip_file: str, file: str):
...
-            files_to_zip.append(file)
 
-    return files_to_zip
+def _to_be_zipped(file: Path) -> bool:
+    if file.is_dir():
+        return False
 
+    for pattern in IGNORED_FILES:
+        # This has to be a string because of fnmatch
+        if file == pattern or fnmatch.fnmatch(str(file), pattern):
```
Comment (sfc-gh-astus) [inline src/snowcli/zipper.py:59]: `file` have to be changed to string before `for`. Now first part compares `Path` and `str` and second part n times does string from `file`

File: tests/snowpark/test_function.py [hunk 4]
```diff
@@ -19,7 +18,7 @@ def test_deploy_function(
     mock_describe.side_effect = ProgrammingError("does not exist or not authorized")
     ctx = mock_ctx()
     mock_connector.return_value = ctx
-    with project_directory("snowpark_functions"):
+    with project_directory("snowpark_functions") as tmp:
```
Comment (sfc-gh-astus) [inline tests/snowpark/test_function.py:21]: `tmp` says nothing, maybe `poject_dir`?

File: tests/test_zipper.py [hunk 5]
```diff
@@ -61,14 +61,14 @@ def test_replace_handler_in_zip(temp_dir, app_zip):
...
-        zip_file_path=app_zip,
+        zip_file_path=str(app_zip),
         coverage_reports_stage="@example",
         coverage_reports_stage_path="test_db.public.example",
     )
     assert os.path.isfile(app_zip)
     assert result == "snowpark_coverage.measure_coverage"
 
-    with ZipFile(app_zip, "r") as zip:
+    with ZipFile(str(app_zip), "r") as zip:
```
Comment (sfc-gh-astus) [inline tests/test_zipper.py:71]: Why not `Path`?

File: src/snowcli/zipper.py (hunk 3 above)
Comment (sfc-gh-turbaszek) [reply src/snowcli/zipper.py:59]: Good point, fixed

File: src/snowcli/cli/snowpark/commands.py [hunk 6]
```diff
@@ -122,37 +125,76 @@ def deploy(
...
+
+        operation_result = _deploy_single_object(
+            manager=pm,
+            object_type=ObjectType.PROCEDURE,
+            object_definition=procedure,
+            replace=replace,
+            packages=packages,
+            stage_artifact_path=artifact_stage_target,
+        )
+        return CollectionResult([operation_result])
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowcli/cli/snowpark/commands.py:179]: Very good point 🚀

File: tests/test_zipper.py (hunk 5 above)
Comment (sfc-gh-turbaszek) [reply tests/test_zipper.py:71]: `Expected type 'Union[str, PathLike[str]]', got 'Generator' instead` warning in IDE, string solves the issue

File: src/snowcli/cli/snowpark/commands.py [hunk 7]
```diff
@@ -122,37 +126,60 @@ def deploy(
             "Using coverage wrapper is currently limited to project with single procedure."
         )
 
-    build_artifact_path = Path("app.zip")
+    build_artifact_path = _get_snowpark_artefact_path(snowpark)
+    zip_package = build_artifact_path.stem
```
Comment (sfc-gh-pjob) [inline src/snowcli/cli/snowpark/commands.py:130]: It looks like this variable is unused

File: src/snowcli/cli/snowpark/commands.py [hunk 8]
```diff
@@ -122,37 +126,60 @@ def deploy(
...
+
     # TODO: this should be configurable
```
Comment (sfc-gh-pjob) [inline src/snowcli/cli/snowpark/commands.py:132]: Is this TODO still valid?

File: src/snowcli/cli/snowpark/commands.py [hunk 9]
```diff
@@ -122,37 +126,60 @@ def deploy(
...
+            build_artifact_path=build_artifact_path,
+            packages=packages,
+            pm=pm,
+            procedure=procedures[0],
+            replace=replace,
+            stage_manager=stage_manager,
+            stage_name=stage_name,
+        )
 
     # TODO: Check if any object already exists before we start updates
```
Comment (sfc-gh-pjob) [inline src/snowcli/cli/snowpark/commands.py:166]: Do we have any task for it? Is it still valid?

File: src/snowcli/cli/snowpark/procedure_coverage/manager.py [hunk 10]
```diff
@@ -48,15 +44,17 @@ def report(
         orig_get_python_source = coverage.python.get_python_source
 
         def new_get_python_source(filename: str):
-            new_filename = filename[filename.index("app.zip/") + len("app.zip/") :]
-            return orig_get_python_source(new_filename)
+            file_path = Path(filename)
+            parts = file_path.parts
+            new_path = Path(artefact_name) / "/".join(
+                parts[parts.index("app.zip") + 1 :]
```
Comment (sfc-gh-pjob) [inline src/snowcli/cli/snowpark/procedure_coverage/manager.py:50]: Is `app.zip` present here a different `app.zip` than our previous hardcoded artefact name?

File: src/snowcli/cli/snowpark/procedure_coverage/commands.py [hunk 11]
```diff
@@ -42,10 +46,13 @@ def procedure_coverage_report(
     store_as_comment: bool = StoreAsCommandOption,
     **options,
 ):
+    snowpark = get_snowpark_project_definition()
     message = ProcedureCoverageManager().report(
         identifier=identifier,
         output_format=output_format,
         store_as_comment=store_as_comment,
+        app_stage_path=get_app_stage_path(snowpark),
+        artefact_name="app.zip",
```
Comment (sfc-gh-pjob) [inline src/snowcli/cli/snowpark/procedure_coverage/commands.py:55]: I don't think I understand why app.zip is hardcoded here?

File: tests_integration/test_procedure.py [hunk 12]
```diff
@@ -66,11 +69,13 @@ def test_snowpark_procedure_flow(
             f"{function_name}(VARCHAR) RETURN VARCHAR"
         )
 
+        # TODO: this should be unique to avoid side effects between parallel runs
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply tests_integration/test_procedure.py:72]: Good point, I forgot the stage will be created within database namespace

File: src/snowcli/cli/snowpark/commands.py (hunk 9 above)
Comment (sfc-gh-turbaszek) [reply src/snowcli/cli/snowpark/commands.py:166]: Yes, I created a task for it

File: src/snowcli/cli/snowpark/procedure_coverage/manager.py (hunk 10 above)
Comment (sfc-gh-turbaszek) [reply src/snowcli/cli/snowpark/procedure_coverage/manager.py:50]: Good catch, fixed

File: src/snowcli/cli/snowpark/procedure_coverage/commands.py (hunk 11 above)
Comment (sfc-gh-turbaszek) [reply src/snowcli/cli/snowpark/procedure_coverage/commands.py:55]: Fixed

**Review by sfc-gh-pjob — APPROVED** [review]

**Pull request merged**